go/ql/lib/semmle/go/frameworks/SQL.qll

```diff
@@ -145,9 +145,25 @@ module SQL {
           f.hasQualifiedName(gopgorm(), "Q") and
           arg = 0
           or
-          exists(string tp, string m | f.(Method).hasQualifiedName(gopgorm(), tp, m) |
+          exists(string tp, string m | f.(Method).hasQualifiedName([gopgorm(), gopg()], tp, m) |
+            tp = ["DB", "Conn"] and
+            m = ["QueryContext", "QueryOneContext"] and
+            arg = 2
+            or
+            tp = ["DB", "Conn"] and
+            m = ["ExecContext", "ExecOneContext", "PrepareContext", "Query", "QueryOne"] and
```

Comment:
I couldn't find `PrepareContext` in the documentation.

Reply:
Will remove, thanks

Reply:
I think you forgot to do this.

Reply:
Sorry, should be removed now

Reply:
@owen-mc BTW, I can see we have sinks for interface methods such as `database.DB.ExecContext` but they are not applied to implementing methods (eg: `github.com/go-pg/DB.ExecContext`). I guess this is because the way of class "inheritance" works in Go, but if a class implements all methods defined in an interface, can we consider that it implements it and apply the models we have for the interface?

Reply:
Hm, it looks like we have a model [here](https://github.com/github/codeql/blob/1767ca268201178308383fd793ce5ca8dd80cdbd/go/ql/lib/semmle/go/frameworks/stdlib/DatabaseSql.qll#L62) and [here](https://github.com/github/codeql/blob/1767ca268201178308383fd793ce5ca8dd80cdbd/go/ql/lib/semmle/go/frameworks/stdlib/DatabaseSql.qll#L95) for [ExecerContext.ExecContext](https://pkg.go.dev/database/sql/driver#ExecerContext), which seems to be subtly different from [DB.ExecContext](https://pkg.go.dev/database/sql#DB.ExecContext) (by the type of the third parameter). I think we're actually missing a model for `DB.ExecContext` - maybe whoever did this modelling thought that the models above were enough.

Reply:
To answer your question, if a class implements all methods defined in an interface then a model for the interface method should cover that method in the class. However, we have recently realised that there is an issue if the interface is defined in a library that isn't imported anywhere, because we won't realise that the interface is being implemented.

Reply:
Ok, that explains it, thanks! So in case we add the model for `database.sql.DB.ExecContext` the ones for `go-pg` DB may not be needed. Should I file an issue or try to fix in this PR?

Reply:
Either one is fine. If it isn't much work and you're happy to do it then you can just tack it onto this PR. If you file an issue then what would be really helpful is a list of which functions should be modeled as `SQL::Query::Range`, and which arguments should be modeled as `SQL::QueryString::Range`.

Reply:
@owen-mc I took a look, but seems like `database.sql.DB.ExecContext` is already modelled [here](https://github.com/github/codeql/blob/1767ca268201178308383fd793ce5ca8dd80cdbd/go/ql/lib/semmle/go/frameworks/stdlib/DatabaseSql.qll#L20-L22). The reason we did not report the SQLi in `github.com/go-pg/pg/v10.DB.ExecContext` is because it has a different [signature](https://pkg.go.dev/github.com/go-pg/pg/v10#DB.ExecContext)

```
func (db *DB) ExecContext(ctx context.Context, query string, args ...any) (Result, error)
```

vs

```
func (db DB) ExecContext(c context.Context, query interface{}, params ...interface{}) (Result, error)
```

Since the GO-PG ones have already been modelled, I think we are good

Reply:
Ah, I missed that in my searching. Well spotted. Yes, that all seems good then.